### PR TITLE
[SPARK-49191][SS] Add support for reading transformWithState map state variables with state data source reader

### PR DESCRIPTION
**Pull request by jingz-db**
### What changes were proposed in this pull request?

Add support for reading transformWithState map state variables with state data source reader.

### Why are the changes needed?

Changes are needed to integrate reading state reading with new operator metadata and state schema format for the map state types used in state variables within transformWithState

### Does this PR introduce _any_ user-facing change?

No. Similar way as reading valueState, user can now read mapState state var as:
```
spark
   .read
   .format("statestore")
   .option("operatorId", <operatorId>)
   .option("stateVarName", <mapStateVarName>)
   .load(<state path>)
```
The output Dataframe will look like:
```
+----+---------------------------+------------+
|key |map_value                  |partition_id|
+----+---------------------------+------------+
|{k1}|{{v2} -> {5}, {v1} -> {10}}|0           |
|{k2}|{{v2} -> {3}}              |0           |
+----+---------------------------+------------+
```
Or this if TTL is enabled:
```
+----+------------------------------------------------+------------+
|key |map_value                                       |partition_id|
+----+------------------------------------------------+------------+
|{k1}|{{key2} -> {{2}, 61000}, {key1} -> {{1}, 61000}}|0           |
+----+------------------------------------------------+------------+
```
An example schema for output dataframe:
```
root
 |-- key: struct (nullable = true) # grouping key row
 |    |-- value: string (nullable = true)
 |-- map_value: map (nullable = true) 
 |    |-- key: struct # user key row
 |    |    |-- value: string (nullable = true)
 |    |-- value: struct (valueContainsNull = false) # value row in state store
 |    |    |-- value: struct (nullable = true) # value row
 |    |    |    |-- value: integer (nullable = false)
 |    |    |-- ttlExpirationMs: long (nullable = true) # ttl column
 |-- partition_id: integer (nullable = true)
```
### How was this patch tested?

Unit tests in `StateDataSourceTransformWithStateSuite`

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StatePartitionReader.scala
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StatePartitionReader.scala:71]: I guess this portion might change a little bit after the list state related PR is merged ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StatePartitionReader.scala
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StatePartitionReader.scala:172]: No change here right ? just moved under the else block ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/utils/SchemaUtil.scala
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/utils/SchemaUtil.scala:274]: Is this required ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/utils/SchemaUtil.scala
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/utils/SchemaUtil.scala:287]: nit: for multi-line comments - I guess we should start from the next line ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/utils/SchemaUtil.scala
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/utils/SchemaUtil.scala:302]: nit: same as above

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/utils/SchemaUtil.scala
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/utils/SchemaUtil.scala:320]: nit: same as above

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/utils/SchemaUtil.scala
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/utils/SchemaUtil.scala:137]: Probably not globally sorted. but we could say that all entries with the same grouping key should be co-located ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/utils/SchemaUtil.scala
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/utils/SchemaUtil.scala:99]: nit: same here

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/utils/SchemaUtil.scala
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/utils/SchemaUtil.scala:128]: So this mapping is just based on index ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StatePartitionReader.scala
Comment (jingz-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StatePartitionReader.scala:71]: The keySchema should be the same (we cannot use dummy schema on key row), but yes, the valueSchema will change a bit after list state PR is merged.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StatePartitionReader.scala
Comment (jingz-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StatePartitionReader.scala:172]: Yeah, no change.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/utils/SchemaUtil.scala
Comment (jingz-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/utils/SchemaUtil.scala:128]: Yeah, the mapping is based on index. The constructor will do something like this:
```
  def apply(keys: Array[_], values: Array[_]): ArrayBasedMapData = {
    new ArrayBasedMapData(new GenericArrayData(keys), new GenericArrayData(values))
  }
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/utils/SchemaUtil.scala
Comment (jingz-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/utils/SchemaUtil.scala:274]: Modified as `MapType.apply()` to get rid of assigning `valueContainsNull = false`.

**Review by anishshri-db — APPROVED** [review]
lgtm, thanks !

**Comment by HyukjinKwon** [comment]
cc @HeartSaVioR

**Comment by HeartSaVioR** [comment]
I also commented in https://github.com/apache/spark/pull/47978 - looks like same issue appears here. The list state and map state are meant to be scalable, flipping the coin, they could be beyond the memory if we materialize all elements. Could you please follow/join the discussion in #47978 and reflect the decision once we have a decision? Thanks!

**Comment by HeartSaVioR** [comment]
Please ignore above comment. We will proceed with the current proposed UX.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/utils/SchemaUtil.scala
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/utils/SchemaUtil.scala:311]: I'd rather not leave this to raise NoSuchElementException which would not be helpful for both users and developers. Shall we explicitly throw an internal exception with message containing context of this error?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/utils/SchemaUtil.scala
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/utils/SchemaUtil.scala:126]: ArrayBasedMapData has an apply method which directly receives Map (though "immutable" Map) and handles the underlying thing. I guess this may get benefit by avoiding conversion from mutable Map to immutable one, but it sounds to me as a premature optimization.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/utils/SchemaUtil.scala
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/utils/SchemaUtil.scala:146]: nit: shall we name this better to be self-describe? e.g. `foundNewGroupingKey` or so. I know it doesn't fit with the case of first grouping key, but except the case (and it's only exceptional in this loop and no longer be an exception outside the loop), it's better to understand about the meaning of the variable.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/utils/SchemaUtil.scala
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/utils/SchemaUtil.scala:168]: This code comment justifies that the variable name needs to reflect this.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/utils/SchemaUtil.scala
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/utils/SchemaUtil.scala:121]: nit: `construct` or `create` is probably clearer as this doesn't update an existing row but creates a new row.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/utils/SchemaUtil.scala
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/utils/SchemaUtil.scala:181]: This shouldn't happen if caller checks hasNext before, right? There is no case where null as InternalRow is expected.

Let's follow Iterator's contract. We should throw NoSuchElementException, or if we want to track this separately, use Spark's internal exception.

**Review by HeartSaVioR — COMMENTED** [review]
First pass. Looks OK in overall. Minors.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/utils/SchemaUtil.scala
Comment (jingz-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/utils/SchemaUtil.scala]: Sure, that makes a lot of sense. I modified this to be throw `NoSuchElementException` the same as other iterators.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/utils/SchemaUtil.scala
Comment (jingz-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/utils/SchemaUtil.scala]: Thanks! Modified to call apply on the map directly.

**Review by HeartSaVioR — APPROVED** [review]
+1

**Comment by HeartSaVioR** [comment]
Thanks! Merging to master.

**Pull request closed without merging**
